geoh5py/objects/surveys/electromagnetics/base.py

```diff
@@ -219,55 +231,61 @@ def copy(  # pylint: disable=too-many-arguments
         if parent is None:
             parent = self.parent
 
-        omit_list = [
-            "_metadata",
-            "_receivers",
-            "_transmitters",
-            "_base_stations",
-            "_tx_id_property",
-        ]
-        kwargs["omit_list"] = omit_list
-        metadata = self.metadata.copy()
         new_entity = super().copy(
             parent=parent,
             clear_cache=clear_cache,
             copy_children=copy_children,
             mask=mask,
             cell_mask=cell_mask,
+            omit_list=OMIT_LIST,
             **kwargs,
         )
 
-        metadata["EM Dataset"][new_entity.type] = new_entity.uid
+        for key, value in self.metadata["EM Dataset"].items():
+            if not isinstance(value, (uuid.UUID, type(None))):
+                new_entity.edit_metadata({key: value})
 
-        if apply_to_complement and (getattr(self, "complement", None) is not None):
-            base_object = (
-                self.base_transmitter_type  # pylint: disable=no-member
-                if isinstance(self, self.default_receiver_type)
-                else self.base_receiver_type  # pylint: disable=no-member
-            )
-            new_complement = super(  # pylint: disable=bad-super-call
-                base_object, self.complement  # pylint: disable=no-member
-            ).copy(
+        if self.complement is not None:
+            self.copy_complement(
+                new_entity,
                 parent=parent,
-                omit_list=omit_list,
                 copy_children=copy_children,
                 clear_cache=clear_cache,
                 mask=mask,
             )
 
-            setattr(
-                new_entity,
-                self.complement.type,  # pylint: disable=no-member
-                new_complement,
-            )
-            metadata["EM Dataset"][
-                self.complement.type  # pylint: disable=no-member
-            ] = new_complement.uid
-            new_complement.metadata = metadata
-
-        new_entity.metadata = metadata
         return new_entity
 
+    def copy_complement(
+        self,
+        new_entity,
+        parent=None,
```

Comment:
**suggestion:** add type hints for params and return type
possibly also some doc

Reply:
Done